portality/crosswalks/article_crossref_xml.py

```diff
@@ -171,13 +172,37 @@ def crosswalk_article(self, record, journal):
         md = journal.find("x:journal_metadata", NS)
         if md is not None:
             issns = md.findall("x:issn", NS)
-            if issns is not None:
-                for issn in issns:
-                    if issn.attrib["media_type"] is None or issn.attrib["media_type"] == 'print':
-                        bibjson.add_identifier(bibjson.P_ISSN, issn.text.upper())
-                    elif issn.attrib["media_type"] == 'electronic':
-                        bibjson.add_identifier(bibjson.E_ISSN, issn.text.upper())
 
+            if len(issns) == 1:
+                if len(issns[0].attrib) == 0 or issns[0].attrib["media_type"] == 'electronic':
+                    bibjson.add_identifier(bibjson.E_ISSN, issns[0].text.upper())
+                elif issns[0].attrib["media_type"] == 'print':
+                     bibjson.add_identifier(bibjson.P_ISSN, issns[0].text.upper())
+
+            elif len(issns) == 2:
+                attrs = [0, 0]
+                if len(issns[0].attrib) != 0:
+                    attrs[0] = issns[0].attrib["media_type"]
+                if len(issns[1].attrib) != 0:
+                    attrs[1] = issns[1].attrib["media_type"]
+
+                if bool(attrs[0]) != bool(attrs[1]):
+                    if attrs[0] != 0:
+                        if attrs[0] == "electronic":
+                            attrs[1] = "print"
+                        else:
+                            attrs[1] = "electronic"
+                    else:
+                        if attrs[1] == "electronic":
+                            attrs[0] = "print"
+                        else:
+                            attrs[0] = "electronic"
+                elif attrs[0] == 0:
+                    attrs[0] = "electronic"
+                    attrs[1] = "print"
+
+                bibjson.add_identifier(bibjson.P_ISSN if attrs[0] == "print" else bibjson.E_ISSN, issns[0].text.upper())
+                bibjson.add_identifier(bibjson.P_ISSN if attrs[1] == "print" else bibjson.E_ISSN, issns[1].text.upper())
```

Comment:
A couple more thoughts here about error resilience.  What happens if there are 3 issns (I'm not sure what the schema allows), and what happens if both issns have the same media_type (like, if they both say they are print issns?).

I tried to imagine a bit of code that handled all the cases, and I haven't tested it but something like this might work:

```
unalloc = []
pissns = []
eissns = []

for issn in issns:
    if issn.attrib["media_type"] == 'print':
       pissns.append(issn)
    elif issn.attrib["media_type"] == 'electronic':
       eissns.append(issn)
    else:
        unalloc.append(issn)

if len(pissn) > 1 or len(eissn) > 1: raise Exception("too many issns or issns of the same type")

for issn in unalloc:
    if len(eissn) == 0: 
        eissn.append(issn)
        continue
     if len(pissn) == 0:
        pissn.append(issn)
        continue
      raise Exception("too many issns)

# now eissn has 1 issn, and pissn has 0 or more issns, so they can be added using add_identifier if present
```

Reply:
That's a very good point! I've just checked that and Crossref schema allows max 6(!) ISSNs! But I am wondering why? Are there really cases of journals with 6 valid ISSNs? On the other hand: our schema does not limit the max number of ISSNs at all. If we want to introduce the error like you suggest it is (again) the case of more strict schema then "original" crossref one. I suspect we should also inform the users... and Crossref? Other solution would be changed to our data model and introduction of ISSN arrays...

On the third hand, I do not think we ever had such a problem. Maybe it would be wise to leave it alone for now?

Reply:
I'm opening a new issue for this. I would suggest deploying this change and solving that problem separately.
https://github.com/DOAJ/doajPM/issues/2806

Reply:
> On the third hand, I do not think we ever had such a problem. Maybe it would be wise to leave it alone for now?

Before we do that we should have a hard think about what the possible consequences are.  Cleaning up broken article data is hard and time consuming so I'd err on the side of caution.

Let me make sure I've got this right - your code above only works if there are 1 or 2 issns.  If there are more than that, does the article simply fail to import?

Reply:
I do agree with you it should be handled somehow. I just think we could make it in the separate issue to not delay the deployment of this issue. Especially that it does not "break" the too-many-issns-handling, it just does not fix it (situation is the same as before this issue has been raised). I can pick that up as soon as we've got a decision from Dom - implementation shouldn't be difficult,

>Let me make sure I've got this right - your code above only works if there are 1 or 2 ISSNs. If there are more than that, does the article simply fail to import?

I explained that in the created issue and I added the additional XML upload file. If there is more then 2 issns the last two are taken into consideration.

Reply:
Journals can't have 6 ISSNs! That's madness. There are only three types of ISSN: P-ISSN, E-ISSN and ISSN-L.

We haven't adopted ISSN-L because its uptake was low.